api/src/main/java/org/eclipse/microprofile/config/spi/ConfigSourceProvider.java

```diff
@@ -44,8 +44,11 @@
  */
 public interface ConfigSourceProvider {
     /**
-     * Return the {@link ConfigSource} instances that are provided by this provider. An empty {@code Iterable} may
-     * be returned if no sources are to be provided.
+     * Return the {@link ConfigSource} instances that are provided by this provider.  
+     * The {@link java.util.Iterator Iterable} contains a fixed number of {@linkplain ConfigSource configuration sources}, 
+     * determined at application start time, and the config sources themselves may be static or dynamic.
```

Comment:
How does the `ConfigSourceProvider` know when "application start time" is?

Reply:
I've also stated that adding a restriction on the sources determined at start time, removes the ability to register additional sources later on. For instance, `SmallRyeConfig` provides an API to do just that.

I'm not against this restriction, but then I propose that the spec is also updated to clarify that sources may only be discovered on Config initialization.

Reply:
> How does the `ConfigSourceProvider` know when "application start time" is?

I think the most correct timeline is on `Config` initialization. Remember that MP Config may be used outside of application containers, where the notion of "application start time" is not well defined.

Even on application container, due to the programmatic API, nothings stops you to access the Config before the application starts, so you need to have the Config instance ready.

Reply:
I don't understand the focus on lifecycle concerns.  Isn't this proposed restriction just saying: if the method returns sources A, B and C the first time it is invoked, then it must return sources A, B and C the second time it is invoked, and the third, and the nth?  Or, much more clearly and succinctly: isn't this just saying the method must be deterministic?

Reply:
It is what it says, yes.

My point is, once you get a `Config` instance, you are locked in the registered `ConfigSources`. Right now, the spec did not specify that you couldn't add additional sources after getting a hold of the `Config`. With this change it effectively locks the `Config`. Is this a reasonable assumption?

If so, I'm only requesting that we clarify in the actual spec documentation the proper behaviour, instead of just being in the java docs. This seems to be somehow important to only be referenced in the java docs.

Reply:
added more clarification on the spec.